File: tools/attention_tokenizer.py
# ============================================================
# Task 2 (tiny-attention-spelling-reranker-FIXED plan) — Python mirror of
# src/attention-tokenizer.mjs. ONE immutable ID contract:
#   0 PAD  1 UNK  2 BOS  3 EOS  4 TARGET  5 PROTECTED  6 PUNCT  7 MASK
#   8..8191 learned vocabulary
#
# Model input is NFC-normalized + Unicode-lowercased ONLY here; Vietnamese
# diacritics are NEVER stripped. Character n-grams are Unicode 2..4-grams
# hashed with FNV-1a over UTF-8 bytes (unsigned 32-bit) modulo 4096.
#
# Units (surface/kind/start/end) come from the JavaScript ValidationDocument
# pipeline (exported in golden fixtures / ranking rows) — production Python
# never re-tokenizes raw SMS text. This module owns ids, markers, context
# selection and char hashing, byte-identically to the JS reference.
# ============================================================
from __future__ import annotations

import unicodedata
from typing import Dict, List, Optional, Sequence, Tuple

ATTENTION_TOKENIZER_VERSION = "attention-tokenizer-v1"

SPECIAL_IDS = {
    "PAD": 0, "UNK": 1, "BOS": 2, "EOS": 3,
    "TARGET": 4, "PROTECTED": 5, "PUNCT": 6, "MASK": 7,
}
MARKER_NONE = 0
MARKER_TARGET = 1
MARKER_PROTECTED = 2
MARKER_PUNCT = 3

VOCAB_SIZE = 8192
FIRST_LEARNED_ID = 8
CHAR_HASH_BUCKETS = 4096
CHAR_NGRAM_MIN = 2
CHAR_NGRAM_MAX = 4
MAX_CHAR_NGRAMS = 32
MAX_CONTEXT_TOKENS = 32

_FNV_OFFSET_BASIS = 0x811C9DC5
_FNV_PRIME = 0x01000193
_MASK32 = 0xFFFFFFFF


def fnv1a32_utf8(text: str) -> int:
    """FNV-1a, 32-bit, over UTF-8 bytes."""
    h = _FNV_OFFSET_BASIS
    for byte in text.encode("utf-8"):
        h ^= byte
        h = (h * _FNV_PRIME) & _MASK32
    return h


def hash_char_ngram(ngram: str) -> int:
    return fnv1a32_utf8(ngram) % CHAR_HASH_BUCKETS


def normalize_for_model(text: str) -> str:
    """NFC + Unicode lowercase; accents preserved."""
    return unicodedata.normalize("NFC", text).lower()


def char_ngram_hashes(word: str) -> List[int]:
    """Ordered, de-duplicated Unicode character 2..4-gram hashes."""
    s = normalize_for_model(word)
    seen = set()
    out: List[int] = []
    for length in range(CHAR_NGRAM_MIN, CHAR_NGRAM_MAX + 1):
        for i in range(0, len(s) - length + 1):
            g = s[i:i + length]
            if g not in seen:
                seen.add(g)
                out.append(hash_char_ngram(g))
    return out[:MAX_CHAR_NGRAMS]


def select_context_indices(
    n: int, target_idx: int, max_tokens: int = MAX_CONTEXT_TOKENS,
) -> List[int]:
    """<=max keep all; else first four + target + nearest by
    (absoluteDistance, index); result sorted ascending."""
    if n <= max_tokens:
        return list(range(n))
    if not 0 <= target_idx < n:
        raise ValueError("target out of range")
    kept = {0, 1, 2, 3, target_idx}
    rest = [i for i in range(n) if i not in kept]
    rest.sort(key=lambda i: (abs(i - target_idx), i))
    for i in rest:
        if len(kept) >= max_tokens:
            break
        kept.add(i)
    return sorted(kept)


def encode_context_units(
    units: Sequence[dict],
    target_idx: int,
    vocab: Optional[Dict[str, int]] = None,
) -> dict:
    """Encode one target-in-context example. Returns a dict whose JSON form
    is identical to the JavaScript reference implementation's output."""
    selected = select_context_indices(len(units), target_idx)
    ids: List[int] = []
    markers: List[int] = []
    char_hashes: List[List[int]] = []
    mask: List[int] = []
    target_position = -1
    for slot, unit_idx in enumerate(selected):
        is_target = unit_idx == target_idx
        unit = units[unit_idx]
        kind = unit["kind"]
        if kind == "protected":
            uid, marker, hashes = SPECIAL_IDS["PROTECTED"], MARKER_PROTECTED, []
        elif kind == "punct":
            uid, marker, hashes = SPECIAL_IDS["PUNCT"], MARKER_PUNCT, []
        else:
            norm = normalize_for_model(unit["surface"])
            uid = vocab.get(norm, SPECIAL_IDS["UNK"]) if vocab else SPECIAL_IDS["UNK"]
            marker = MARKER_TARGET if is_target else MARKER_NONE
            # Hide the surface under correction; the TARGET marker identifies
            # the slot without leaking the misspelling into the encoder.
            if is_target:
                uid, hashes = SPECIAL_IDS["MASK"], []
            else:
                hashes = char_ngram_hashes(norm)
        if is_target:
            target_position = slot
        ids.append(uid)
        markers.append(marker)
        char_hashes.append(hashes)
        mask.append(1)
    if target_position < 0:
        raise ValueError("target not selected")
    return {
        "ids": ids,
        "markers": markers,
        "charHashes": char_hashes,
        "mask": mask,
        "selectedIndices": selected,
        "targetPosition": target_position,
    }


def encode_option_surface(
    surface: str, vocab: Optional[Dict[str, int]] = None,
) -> dict:
    """Pure per-surface option encoding (KEEP_ORIGINAL included)."""
    norm = normalize_for_model(surface)
    uid = vocab.get(norm, SPECIAL_IDS["UNK"]) if vocab else SPECIAL_IDS["UNK"]
    return {"id": uid, "charHashes": char_ngram_hashes(norm)}

File: src/attention-tokenizer.mjs
// ============================================================
// Task 2 (tiny-attention-spelling-reranker-FIXED plan) — shared attention
// tokenizer + 32-token context selector.
//
// ONE immutable ID contract for JavaScript and Python:
//   0 PAD  1 UNK  2 BOS  3 EOS  4 TARGET  5 PROTECTED  6 PUNCT  7 MASK
//   8..8191 learned vocabulary
//
// Model input is NFC-normalized + Unicode-lowercased ONLY here; Vietnamese
// diacritics are NEVER stripped and the original SMS surface/offsets are
// never modified. Character n-grams are Unicode 2..4-grams hashed with
// FNV-1a over UTF-8 bytes (unsigned 32-bit) modulo 4096 — never JS code-unit
// hashing.
//
// Context selection builds on the existing ValidationDocument tokens (no
// re-tokenization): <=32 units keep all; otherwise retain the first four,
// the target, then nearest-by-(absoluteDistance, index) until 32 slots are
// full; retained indices sort ascending; the target slot is marked
// explicitly and is NEVER replaced by a candidate surface.
// ============================================================
import { normalizeNfc } from './normalizer.mjs';

export const ATTENTION_TOKENIZER_VERSION = 'attention-tokenizer-v1';

export const SPECIAL_IDS = Object.freeze({
  PAD: 0, UNK: 1, BOS: 2, EOS: 3, TARGET: 4, PROTECTED: 5, PUNCT: 6, MASK: 7,
});

/** Marker embedding indices (independent namespace from word IDs). */
export const MARKER = Object.freeze({ NONE: 0, TARGET: 1, PROTECTED: 2, PUNCT: 3 });

export const VOCAB_SIZE = 8192;
export const FIRST_LEARNED_ID = 8;
export const CHAR_HASH_BUCKETS = 4096;
export const CHAR_NGRAM_MIN = 2;
export const CHAR_NGRAM_MAX = 4;
export const MAX_CONTEXT_TOKENS = 32;

/** FNV-1a, 32-bit, over UTF-8 bytes. */
export function fnv1a32Utf8(text) {
  let h = 0x811c9dc5;
  for (const byte of new TextEncoder().encode(text)) {
    h ^= byte;
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h >>> 0;
}

export function hashCharNgram(ngram) {
  return fnv1a32Utf8(ngram) % CHAR_HASH_BUCKETS;
}

/** Model-input normalization: NFC + Unicode lowercase (accents kept). */
export function normalizeForModel(text) {
  return normalizeNfc(String(text)).toLowerCase();
}

/** Ordered, de-duplicated character 2..4-grams of a normalized word. */
export function charNgramsOf(word) {
  const seen = new Set();
  const out = [];
  const s = String(word);
  for (let len = CHAR_NGRAM_MIN; len <= CHAR_NGRAM_MAX; len++) {
    for (let i = 0; i + len <= s.length; i++) {
      const g = s.slice(i, i + len);
      if (!seen.has(g)) {
        seen.add(g);
        out.push(g);
      }
    }
  }
  return out;
}

export function charNgramHashes(word) {
  return charNgramsOf(normalizeForModel(word)).map(hashCharNgram);
}

/**
 * Flatten a ValidationDocument into context units. Consecutive non-
 * whitespace tokens fully inside the SAME protected range merge into ONE
 * protected unit (URLs/phones/placeholders are never decomposed).
 * Unit kinds: 'word' | 'protected' | 'punct' (NUMBER/PUNCTUATION/OTHER).
 * Offsets stay in ORIGINAL-text space.
 */
export function unitsFromDocument(doc) {
  const units = [];
  /** unit under construction for the protected range starting at rs/re */
  let openRangeKey = null;
  for (const t of doc.tokens) {
    if (t.type === 'WHITESPACE') continue;
    const key = doc.insideSingleProtectedRange(t.start, t.end)
      ? (() => {
        const r = doc.protectedRanges.find((rr) => rr.contains(t.start, t.end));
        return `${r.start}:${r.end}`;
      })()
      : null;

    if (key !== null && key === openRangeKey && units.length > 0) {
      // extend the currently open protected unit across this token
      const last = units[units.length - 1];
      last.end = t.end;
      last.surface = doc.originalText.slice(last.start, last.end);
      continue;
    }
    openRangeKey = key;

    if (key !== null) {
      units.push({
        kind: 'protected',
        surface: doc.originalText.slice(t.start, t.end),
        start: t.start, end: t.end,
      });
      continue;
    }
    units.push({
      kind: t.type === 'WORD' ? 'word' : 'punct',
      surface: doc.originalText.slice(t.start, t.end),
      start: t.start, end: t.end,
    });
  }
  return units.map(Object.freeze);
}

/**
 * Deterministic context selection. Returns ascending retained unit indices.
 */
export function selectContextIndices(n, targetIdx, max = MAX_CONTEXT_TOKENS) {
  if (n <= max) return Array.from({ length: n }, (_, i) => i);
  if (targetIdx < 0 || targetIdx >= n) throw new RangeError('target out of range');
  const kept = new Set([0, 1, 2, 3, targetIdx]);
  const rest = [];
  for (let i = 0; i < n; i++) {
    if (!kept.has(i)) rest.push(i);
  }
  rest.sort((a, b) => Math.abs(a - targetIdx) - Math.abs(b - targetIdx) || a - b);
  for (const i of rest) {
    if (kept.size >= max) break;
    kept.add(i);
  }
  return [...kept].sort((a, b) => a - b);
}

function unitIdAndMarker(unit, isTarget, vocabMap) {
  if (unit.kind === 'protected') {
    return { id: SPECIAL_IDS.PROTECTED, marker: MARKER.PROTECTED, hashes: [] };
  }
  if (unit.kind === 'punct') {
    return { id: SPECIAL_IDS.PUNCT, marker: MARKER.PUNCT, hashes: [] };
  }
  const norm = normalizeForModel(unit.surface);
  const id = vocabMap?.get(norm) ?? SPECIAL_IDS.UNK;
  return {
    id,
    marker: isTarget ? MARKER.TARGET : MARKER.NONE,
    hashes: charNgramHashes(norm),
  };
}

/**
 * Encode one target-in-context example into model vectors.
 * @returns {{ids:number[], markers:number[], charHashes:number[][],
 *            mask:number[], selectedIndices:number[], targetPosition:number}}
 */
export function encodeContextUnits(units, targetIdx, vocabMap) {
  const selectedIndices = selectContextIndices(units.length, targetIdx);
  const ids = [];
  const markers = [];
  const charHashes = [];
  const mask = [];
  let targetPosition = -1;
  selectedIndices.forEach((unitIdx, slot) => {
    const isTarget = unitIdx === targetIdx;
    if (isTarget) targetPosition = slot;
    const { id, marker, hashes } = unitIdAndMarker(units[unitIdx], isTarget, vocabMap);
    ids.push(id);
    markers.push(marker);
    charHashes.push(hashes);
    mask.push(1);
  });
  if (targetPosition < 0) throw new RangeError('target not selected');
  return { ids, markers, charHashes, mask, selectedIndices, targetPosition };
}

/**
 * Encode an option surface (KEEP_ORIGINAL uses the original token's
 * normalized form). Pure per-surface function: candidate list order can
 * never influence option vectors.
 */
export function encodeOptionSurface(surface, vocabMap) {
  const norm = normalizeForModel(surface);
  return {
    id: vocabMap?.get(norm) ?? SPECIAL_IDS.UNK,
    charHashes: charNgramHashes(norm),
  };
}
